lighthouse-assert/assert/difference/object.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import { IDifference } from './interface';
+import { INoneObjectActualAudit, INoneObjectExpectedAudit, IDiff, INormalizedExpectedScore } from '../types';
+import { inRange } from '../utils';
+
+const OPERAND_EXPECTATION_REGEXP = /^(<=?|>=?)/;
+
+export class ObjectDifference implements IDifference {
+  private path: string;
+  private actual: INoneObjectActualAudit;
+  private expected: INoneObjectExpectedAudit;
+
+  /**
+   * Constructor
+   * @param {string} path
+   * @param {IActualAudit} actual
+   * @param {IExpectedAudit} expected
+   */
+  constructor(path: string, actual: INoneObjectActualAudit, expected: INoneObjectExpectedAudit) {
+    this.path = path;
+    this.actual = actual;
+    this.expected = expected;
+  }
+
+  /**
+   * Find diff
+   *
+   * @return {IDiff}
+   */
+  getDiff(): IDiff {
+    let diff: IDiff = {};
+
+    if (this.matchesExpectation()) return diff;
+    for (const expectationType of Object.keys(this.expected)) {
+      const expectedByType = this.expected[expectationType];
+      // If they aren't both an object we can't recurse further, so this is the difference.
+      if (this.actual.score === null || expectedByType === null ||
+        typeof this.actual.score !== 'object' || typeof expectedByType !== 'object' ||
+        expectedByType instanceof RegExp) {
+        diff = {path: this.path, actual: this.actual.score, expected: this.expected.score};
+      }
+    }
+    return diff;
+  }
+
+  /**
+   * Checks if the actual value matches the expectation. Does not recursively search. This supports
+   * greater than/less than operators, e.g. "<100", ">90"
+   *
+   * @return {boolean}
+   */
+  matchesExpectation(): boolean {
+    const actualValue = this.normalize(this.actual.score);
+    const normalizedExpected: INormalizedExpectedScore = {
+      warn: this.normalize(this.expected.score.warn),
+      error: this.normalize(this.expected.score.error)
+    };
+    // @todo check inRange due to symbols like '<=', '>=' etc.
+    return inRange(normalizedExpected.error, 100, actualValue);
```

Comment:
what I've been thinking is getting `<=, >=` during normalize, but then understood, why should we allow users to do crazy setups like `score: { warn: '>=70', 'error': '<70'}` or `score: { warn: '<80', error: '>90'}` etc. 
What if we say that they allowed setting expectation lower/equal than result like `score: { warn: '<=70', 'error': '<60'}`, `score: { warn: '70', 'error': '60'}`, '`score: true`', `score: 100`, `score: '<70'`,skippinga cases above?

Reply:
it's useful to have assertions on the length of details, rawValue, etc which was the original motivation for the `>=` I'm not the biggest fan of the warn/error object (what do we do if there's an error property of something we want to check?) 

have we considered other possible approaches for expressing warn vs. error?
what do we think the most common use case is? I'm assuming it's going to be failing this in CI which brings up the question of how valuable warn is and if it'd be ok to have a more roundabout usage for warn. It seems like we'll need to make a decision about whether we want to easily support warn/error and go an eslint-like direction that only asserts score or if we want to easily support assertions on arbitrary properties (like smokehouse does now) and force a separate config if you want warn assertions.

Reply:
Approach having warn and error borrowed from pwmetrics. Why have warn? 
Sometimes lighthouse can return not consistent results for several runs (also other people faced this kind of problem, reff https://medium.com/@trueadm/this-looks-awesome-and-is-something-similar-to-what-i-built-for-react-ebdd77192258). Having warn will just show message and CI pass.

I think we can consider using more than score, and change all `difference` logic to work with abstract thing from inside, e.g. not
`findDifference(auditName, actualResult, expectedResult)` 
but 
`findDifference(auditName, resultValue, expectedResultValue);` 
where `expectedResultValue` -> `{warn: '<=70', error: '<60' } or smth from cases above`. On this inside level we can isolate what this `expectedResultValue` value will be, either `score`, or `rawValue`, or something else (we can have a list of availiable things).

But your idea about config aka eslint can also give us a good opportunity. It can be simpler than current config config aka
```
"first-meaningful-paint": {
  "expected": {
    "warn": "<=85",
    "error": "<=65"
  },
  "compareWith": "rawValue"
},
"redirects-http": {
   "expected": true,
  "compareWith": "score"
},
```

we should choose which approach to use, so I can continue work on that...